internal/verifier/compare.go

```diff
@@ -475,63 +510,166 @@ func (verifier *Verifier) getDocumentsCursor(ctx mongo.SessionContext, collectio
+			}
+
+			for i, el := range aggOptions {
+				if el.Key != "pipeline" {
+					continue
+				}
+
+				aggOptions[i].Value = transformPipelineForToHashedIndexKey(
+					aggOptions[i].Value.(mongo.Pipeline),
+					task,
+				)
+
+				break
+			}
+
+		default:
+			panic("bad doc compare query func: " + verifier.docCompareMethod.QueryFunction())
+		}
+	}
+
+	var cmd bson.D
+
+	switch verifier.docCompareMethod.QueryFunction() {
+	case DocQueryFunctionFind:
+		cmd = append(
+			bson.D{{"find", collection.Name()}},
+			findOptions...,
+		)
+	case DocQueryFunctionAggregate:
+		cmd = append(
+			bson.D{
+				{"aggregate", collection.Name()},
+				{"cursor", bson.D{}},
+			},
+			aggOptions...,
+		)
 	}
+
 	if verifier.readPreference.Mode() != readpref.PrimaryMode {
 		runCommandOptions = runCommandOptions.SetReadPreference(verifier.readPreference)
 		if startAtTs != nil {
+			readConcern := bson.D{
+				{"afterClusterTime", *startAtTs},
+			}
 
 			// We never want to read before the change stream start time,
 			// or for the last generation, the change stream end time.
-			findOptions = append(
-				findOptions,
-				bson.E{"readConcern", bson.D{
-					{"afterClusterTime", *startAtTs},
-				}},
+			cmd = append(
+				cmd,
+				bson.E{"readConcern", readConcern},
 			)
 		}
 	}
-	findCmd := append(bson.D{{"find", collection.Name()}}, findOptions...)
 
 	// Suppress this log for recheck tasks because the list of IDs can be
 	// quite long.
 	if !task.IsRecheck() {
+		extJSON, _ := bson.MarshalExtJSON(cmd, true, false)
+
 		verifier.logger.Debug().
 			Any("task", task.PrimaryKey).
-			Str("findCmd", fmt.Sprintf("%s", findCmd)).
+			Str("cmd", lo.Ternary(
+				extJSON == nil,
+				fmt.Sprintf("%s", cmd),
+				string(extJSON),
+			)).
 			Str("options", fmt.Sprintf("%v", *runCommandOptions)).
-			Msg("getDocuments findCmd.")
+			Msg("getDocuments command.")
 	}
 
-	return collection.Database().RunCommandCursor(ctx, findCmd, runCommandOptions)
+	return collection.Database().RunCommandCursor(ctx, cmd, runCommandOptions)
+}
+
+func transformPipelineForToHashedIndexKey(
+	in mongo.Pipeline,
+	task *VerificationTask,
+) mongo.Pipeline {
+	return append(
+		slices.Clone(in),
+		bson.D{{"$replaceWith", bson.D{
+			// Single-letter field names minimize the document size.
+			{docKeyInCompare, bson.D(lo.Map(
+				task.QueryFilter.GetDocKeyFields(),
+				func(f string, _ int) bson.E {
+					return bson.E{f, "$$ROOT." + f}
+				},
+			))},
+			{"h", bson.D{
+				{"$toHashedIndexKey", bson.D{
+					{"$_internalKeyStringValue", bson.D{
+						{"input", "$$ROOT"},
+					}},
+				}},
+			}},
+			{"s", bson.D{{"$bsonSize", "$$ROOT"}}},
+		}}},
+	)
 }
 
 func (verifier *Verifier) compareOneDocument(srcClientDoc, dstClientDoc bson.Raw, namespace string) ([]VerificationResult, error) {
 	match := bytes.Equal(srcClientDoc, dstClientDoc)
 	if match {
 		return nil, nil
 	}
-	//verifier.logger.Info().Msg("Byte comparison failed for id %s, falling back to field comparison", id)
```

Comment:
Feels like this log line should still be here, just below the hashed comparison early-return block.